lib/wasi/src/runtime/resolver/source.rs

```diff
@@ -0,0 +1,81 @@
+use std::fmt::Debug;
+
+use anyhow::Error;
+use url::Url;
+
+use crate::runtime::resolver::{PackageSpecifier, Summary};
+
+/// Something that packages can be downloaded from.
+#[async_trait::async_trait]
+pub trait Source: Debug {
+    /// An ID that describes this source.
+    fn id(&self) -> SourceId;
+
+    /// Ask this source which packages would satisfy a particular
+    /// [`Dependency`][dep] constraint.
+    ///
+    /// # Assumptions
+    ///
+    /// It is not an error if there are no package versions that may satisfy
+    /// the dependency, even if the [`Source`] doesn't know of a package
+    /// with that name.
+    ///
+    /// A [`Registry`][reg] will typically have a list of [`Source`]s that are
+    /// queried in order. The first [`Source`] to return one or more
+    /// [`Summaries`][Summary] will be treated as the canonical source for
+    /// that [`Dependency`][dep] and no further [`Source`]s will be queried.
+    ///
+    /// [dep]: crate::runtime::resolver::Dependency
+    /// [reg]: crate::runtime::resolver::Registry
+    async fn query(&self, package: &PackageSpecifier) -> Result<Vec<Summary>, Error>;
+}
+
+#[async_trait::async_trait]
+impl<D, S> Source for D
+where
+    D: std::ops::Deref<Target = S> + Debug + Send + Sync,
+    S: Source + Send + Sync + 'static,
+{
+    fn id(&self) -> SourceId {
+        (**self).id()
+    }
+
+    async fn query(&self, package: &PackageSpecifier) -> Result<Vec<Summary>, Error> {
+        (**self).query(package).await
+    }
+}
+
+/// The type of [`Source`] a [`SourceId`] corresponds to.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum SourceKind {
```

Comment:
Yeah, this is what I was referring to when I said I wasn't happy with how `SourceId` and `Source` work. I'm thinking of just removing `SourceId` altogether because it adds issues elsewhere.

I believe the reason Cargo includes a `SourceId` as part of the `PackageId` is because you can use `[patch]` to make the same package version look like it's coming from different places.

Reply:
Let's just remove it completely for now.

We can re-introduce if necessary.

Reply:
Done.